### PR TITLE
Fix/change single asset from my wallet

### PR DESCRIPTION
**Pull request by sssmi**
# Description

Select Single Token from My wallet.

Clicking any token from My Wallet will max out the token input. Additionally, clicking Non pool Token switches the tab to Single Token invest (only for Deep Pools).

## Type of change

- [x] New feature (non-breaking change which adds functionality)

### Deep Pools

http://localhost:8080/#/pool/0xa13a9247ea42d743238089903570127dda72fe4400000000000000000000035d/invest
On invest page after selecting "Single Token" tab, you can select any token from My Wallet on the left.
Selecting a token will change the input token, and max out the input value based on the token balance on wallet.

### wETH pools
http://localhost:8080/#/pool/0x32296969ef14eb0c6d29669c550d4a0449130230000200000000000000000080/invest
Clicking wETH or ETH token will toggle the (w)ETH input token and max out the input value.
Clicking other tokens just maxes out the token input

## Visual context
| Deep pool | Pool with wrapped native asset |
| ---- | ---- |
| Screen Shot 2022-10-28 at 15 07 51 | Screen Shot 2022-10-28 at 15 07 28  |

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have requested at least 2 reviews (If the PR is significant enough, use best judgement here)
- [x] I have commented my code where relevant, particularly in hard-to-understand areas
- [x] If package-lock.json has changes, it was intentional.
- [x] The base of this PR is `master` if hotfix, `develop` if not

### CODE REVIEW

File: src/components/contextual/pages/pool/invest/InvestPage.vue [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+      <BalAccordion
+        v-if="upToLargeBreakpoint"
+        class="mt-4"
+        :sections="[
+          {
+            title: poolSupportsGeneralisedJoin
+              ? `${$t('myWallet2')} ${
+                  hasNativeBalance ? `${nativeBalance} ${nativeCurrency}` : ''
+                }`
+              : $t('poolTransfer.myWalletTokensCard.title'),
```
Comment (garethfuller) [inline src/components/contextual/pages/pool/invest/InvestPage.vue:165]: I'm not a fan of logic like this in the templates, it makes it hard to parse, please extract to a computed property.

File: src/components/contextual/pages/pool/invest/InvestPage.vue [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+const { network } = configService;
+const { pool, loadingPool, transfersAllowed, useNativeAsset } =
+  usePoolTransfers();
+const { isDeepPool, isWethPool } = usePool(pool);
+const { sor, sorReady } = useInvestState();
+const { activeTab } = useInvestPageTabs();
+const { tokenAddresses, amounts } = useInvestState();
+const { upToLargeBreakpoint } = useBreakpoints();
+const { addTokensIn, setAmountsIn, isSingleAssetJoin, amountsIn } =
+  useJoinPool();
```
Comment (garethfuller) [inline src/components/contextual/pages/pool/invest/InvestPage.vue:36]: Does this not break when the pool isn't a deep pool because you don't use the provider: https://github.com/balancer-labs/frontend-v2/pull/2339/files#diff-de8992b55fdef033962c2ab6870a3b4ae1e1f9bba6309a660120f32bc5e82049R25

**Review by garethfuller — CHANGES_REQUESTED** [review]

File: src/components/contextual/pages/pool/invest/InvestPage.vue (hunk 2 above)
Comment (sssmi) [reply src/components/contextual/pages/pool/invest/InvestPage.vue]: It works, but all these properties of `useJoinPool` will be `undefined`.

I moved the logic to it's own component:
`src/components/contextual/pages/pool/invest/InvestPageMyWallet.vue`

File: src/components/contextual/pages/pool/invest/InvestPage.vue (hunk 1 above)
Comment (sssmi) [reply src/components/contextual/pages/pool/invest/InvestPage.vue]: Done

File: src/components/contextual/pages/pool/invest/InvestPage.vue [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+const poolSupportsGeneralisedJoin = computed<boolean>(() => {
+  return !!(pool.value && isDeepPool.value);
+});
+</script>
+
+<template>
+  <div class="invest-page-layout-grid">
+    <div v-if="!upToLargeBreakpoint" class="col-span-5">
+      <InvestPageMyWallet
+        :poolSupportsGeneralisedJoin="poolSupportsGeneralisedJoin"
```
Comment (garethfuller) [inline src/components/contextual/pages/pool/invest/InvestPage.vue:29]: Not a fan of this prop drilling. Why does this need to be passed down as a prop? All these components should have access to the pool state, so in specific cases, it can just check if the pool isDeep.

File: src/components/contextual/pages/pool/invest/InvestPageAccordion.vue [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+      {
+        title: sectionTitle,
+        id: 'myWalletTokens',
+      },
+    ]"
+  >
+    <!-- TODO: Show some 404 message if Pool not found -->
+    <template #myWalletTokens>
+      <InvestPageMyWallet
+        :poolSupportsGeneralisedJoin="poolSupportsGeneralisedJoin"
```
Comment (garethfuller) [inline src/components/contextual/pages/pool/invest/InvestPageAccordion.vue:46]: Again prop drilling here, it's a sign we shouldn't be handling it this way.

File: src/components/contextual/pages/pool/invest/InvestPageInvestBlock.vue [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        <BalTabs
+          v-if="isDeepPool"
+          v-model="activeTab"
+          :tabs="tabs"
+          class="p-0 m-0 -mb-px whitespace-nowrap"
+          noPad
+        />
+      </div>
+    </template>
+    <template v-if="poolSupportsGeneralisedJoin">
```
Comment (garethfuller) [inline src/components/contextual/pages/pool/invest/InvestPageInvestBlock.vue:69]: If we get rid of the prop drilling for this long prop, we can just replace this with isDeep for now.
```suggestion
    <template v-if="isDeepPool">
```

File: src/components/contextual/pages/pool/invest/InvestPageMyWallet.vue [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+
+  if (indexOfAsset >= 0) {
+    amounts.value[indexOfAsset] = maxBalance;
+  }
+}
+
+function handleMyWalletTokenClick(tokenAddress: string, isPoolToken: boolean) {
+  const maxBalance = getMaxBalanceFor(tokenAddress);
+
+  if (props.poolSupportsGeneralisedJoin) {
```
Comment (garethfuller) [inline src/components/contextual/pages/pool/invest/InvestPageMyWallet.vue:81]: Again, you have access to the pool state here, so just isDeep.

**Review by garethfuller — CHANGES_REQUESTED** [review]

File: src/components/_global/BalTabs/BalTabs.vue [hunk 7]
```diff
@@ -33,14 +33,11 @@ export default defineComponent({
   emits: ['selected', 'update:modelValue'],
 
   setup(props, { emit }) {
-    const activeTab = ref(props.modelValue);
```
[the remark this answers is not in the document]
Comment (sssmi) [reply src/components/_global/BalTabs/BalTabs.vue:36]: Yeah but my point is BalTabs don't really need to hold internal state if `activeTab` is just a copy of the prop.

File: src/components/contextual/pages/pool/invest/InvestPageInvestBlock.vue (hunk 5 above)
Comment (sssmi) [reply src/components/contextual/pages/pool/invest/InvestPageInvestBlock.vue]: True. I removed the `poolSupportsGeneralisedJoin` prop

File: src/components/_global/BalTabs/BalTabs.vue (hunk 7 above)
[the remark this answers is not in the document]
Comment (garethfuller) [reply src/components/_global/BalTabs/BalTabs.vue:36]: Yeah fair enough, but did you check for all references to activeTab in the code to make sure it wasn't being used elsewhere?

File: src/components/contextual/pages/pool/invest/InvestPage.vue [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+</script>
+
+<template>
+  <div class="invest-page-layout-grid">
+    <div v-if="!upToLargeBreakpoint" class="col-span-5">
+      <InvestPageMyWallet />
+    </div>
+
+    <div class="col-span-7">
+      <InvestPageInvestBlock></InvestPageInvestBlock>
```
Comment (garethfuller) [inline src/components/contextual/pages/pool/invest/InvestPage.vue:24]:
```suggestion
      <InvestPageInvestBlock />
```

File: src/components/_global/BalTabs/BalTabs.vue (hunk 7 above)
[the remark this answers is not in the document]
Comment (sssmi) [reply src/components/_global/BalTabs/BalTabs.vue:36]: This is just a Vue component, so the state is not shared outside of itself, or am I missing something? 😅

[the remark this answers is not in the document]
Comment (garethfuller) [reply src/components/_global/BalTabs/BalTabs.vue:36]: My bad was thinking about it as if it was a composable.

**Review by garethfuller — APPROVED** [review]
